File: tools/Static_Calibration.m
close all
clear all

%-------------------------------------------------------------------------%
% 
% To obtain Gravitation Field (raw format):
% 1) get the Total Field for your location from here:
%    http://www.ngdc.noaa.gov/geomag-web (tab Magneti3c Field)
%    es. Total Field = 47,241.3 nT
% 2) Convert this values to Gauss (1nT = 10E-5G)
%    es. Total Field = 47,241.3 nT = 0.47241G
% 3) Convert Total Field to Raw value Total Field, which is the
%    Raw Gravitation Field we are searching for
%    Read your magnetometer datasheet and find your gain value,
%    Which should be the same of the collected raw points
%    es. on HMC5883L, given +_ 1.3 Ga as Sensor Field Range settings
%        Gain (LSB/Gauss) = 1090
%        Raw Total Field = Gain * Total Field
%        0.47241 * 1090 = ~515LSB
% 
% Say in Kelowna:
% 1)	The Total Field: 55,092.4 nT
% 2)	Covert this values to Gauss: 55,092.4 nT = 0.551 G
% 3)	Convert Total Field to Raw Value Total Field, essentially the raw Gravitation Field
% a.	On AK8963, given  4,800,000 nT = 48 G
% b.	Gain = 1/[0.15(?T/ LSB)] = 6.67(LSB/?T) = 0.00667(LSB/nT) = 667(LSB/Gauss)
        %(0.6?/LSB is for 14-bit, 0.15?/LSB is for 16-bit, the magnetometer
        %use 16-bit, see MPU9250.h for detals)
% c.	Raw Total Field = Gain * Total Field = 0.551 * 667 = 367.5170LSB
% 4)	The value read from the magnetometer we used is the magnetometer values in milliGauss
%
%-------------------------------------------------------------------------%

mag_raw = 367.5170; % the unit is in LSB, this value is 367.5170 (16-bit, 92 for 14-bit) for Kelowna

% read data obtained from the magnetometer
f_id = fopen('With_Iron_3cm.txt','r'); 
sample = fscanf(f_id,'%f');
data_mG = vec2mat(sample,3); % the unit is in mGauss
gain = 0.667; % the unit is in LSB/mGauss
data_raw = gain .* data_mG;

% write data_raw to text file
f_id2 = fopen('With_Iron_3cm_raw.txt', 'w'); 
for i=1:size(data_raw,1)
   fprintf(f_id2, '%f ', data_raw(i,:));
   fprintf(f_id2, '\n');
end
fclose(f_id2);

% Build the design matrix D
column_num = length(data_raw);

D(1,:) = data_raw(:,1).^2; % x^2
D(2,:) = data_raw(:,2).^2; % y^2
D(3,:) = data_raw(:,3).^2; % z^2
D(4,:) = 2 .* data_raw(:,2) .* data_raw(:,3); % 2yz
D(5,:) = 2 .* data_raw(:,1) .* data_raw(:,3); % 2xz
D(6,:) = 2 .* data_raw(:,1) .* data_raw(:,2); % 2xy
D(7,:) = 2 .* data_raw(:,1); % 2x
D(8,:) = 2 .* data_raw(:,2); % 2y
D(9,:) = 2 .* data_raw(:,3); % 2z
D(10,:) = ones([1,column_num]); % 1

% Create S matrix
S = D * transpose(D);

% Split the S matrix to 4 sub-matrices S11, S12, S21, and S22
% S21 and transpose(S12)
for m=1:1:6 % row
    for n=1:1:6 %column
        S11(m,n)=S(m,n); % S11
    end
end

for m=1:1:6 
    for n=1:1:4 %column
        S12(m,n)=S(m,n+6); % S12
    end
end  

S21 = transpose(S12); % S21

for m=1:1:4 
    for n=1:1:4 %column
        S22(m,n)=S(m+6,n+6); % S22
    end
end  

% Find the inverse matrix of S22
S22_inv = inv(S22);

% Calculate matrix SS
SS = S11 - S12 * S22_inv * S21;

% Constraint Matrix C
% C = [-1,1,1,0,0,0;1,-1,1,0,0,0;1,1,-1,0,0,0;0,0,0,-4,0,0;0,0,0,0,-4,0;0,0,0,0,0,-4];
% normalized C:
C = [0,0.5,0.5,0,0,0;0.5,0,0.5,0,0,0;0.5,0.5,0,0,0,0;0,0,0,-0.25,0,0;0,0,0,0,-0.25,0;0,0,0,0,0,-0.25];

% Calculate inverse C matrix
C_inv = inv(C);

% Calculate matrix E
E = C * SS;

% Find eigenvalue of matirix E, E*V - V*D = [0]
e = eig(E); % eigenvalue
[V,D] = eig(E); % eigenvector is V

% Find the zero-based position of the only positive eigenvalue. 
% The associated eigenvector will be in the corresponding column of matrix 
maxvalue_e = e(1,1);
for k = 1:1:6 % row of e
    if e(k,1)>0
        maxvalue_e = e(k,1);
        index = k;
        break
    end
end

% Extract the associated eigenvector V1
V1 = V(:,index);

% Check the sign of V1, if it is negative, flip the sign of the eigenvector
if V1(1,1) < 0
    V1 = -1 .* V1;
end

% Calculate V2
V2 = (S22_inv * S21) * V1;

% Setup matrix Ellipsoid, which is what we want!!!
for i = 1:1:6
    for j = 7:1:10
        Ellipsoid(i,1) = V1(i,1);
        Ellipsoid(j,1) = -1 .* V2(j-6,1);
    end
end

% Normalize the Ellipsoid matrix (normalize coefficient of the ellipsoid to prevent trivial results)
Ellipsoid = Ellipsoid ./ norm(Ellipsoid(:));

%-------------------------------------------------------------------------%
% Calculate combined bias
% Setup Q matrix
Q = [Ellipsoid(1,1),Ellipsoid(6,1),Ellipsoid(5,1);Ellipsoid(6,1),Ellipsoid(2,1),Ellipsoid(4,1);Ellipsoid(5,1),Ellipsoid(4,1),Ellipsoid(3,1)];

% Setup U matrix
U = [Ellipsoid(7,1);Ellipsoid(8,1);Ellipsoid(9,1)];

% Calculate inverse Q matrix
Q_inv = inv(Q);

% Calculate B matrix
B = Q_inv * U;

% Combined Bias
x_bias = -B(1,1);
y_bias = -B(2,1);
z_bias = -B(3,1);
bias = repmat([x_bias, y_bias, z_bias],length(data_raw),1); % Hard iron offset

%-------------------------------------------------------------------------%
% Calculate correction matrix A_inv
J = Ellipsoid(10,1);
A_inv = (Q^(0.5)).*((mag_raw)./(sqrt(transpose(B)*Q*B-J)));

%-------------------------------------------------------------------------%
% Calibrated Data
data_cal = transpose(A_inv * transpose(data_raw - bias)); % the matrix dimensions must match 

f_id3 = fopen('With_Iron_3cm_cal.txt', 'w'); % write data_cal to txt file
for i=1:size(data_cal,1)
   fprintf(f_id3, '%f ', data_cal(i,:));
   fprintf(f_id3, '\n');
end
fclose(f_id3);

%-------------------------------------------------------------------------%
% Plot Data
x = data_raw(:,1);
y = data_raw(:,2);
z = data_raw(:,3);
x_cal = data_cal(:,1);
y_cal = data_cal(:,2);
z_cal = data_cal(:,3);

A = Ellipsoid(1,1);
B = Ellipsoid(2,1);
C = Ellipsoid(3,1);
D = Ellipsoid(6,1);
E = Ellipsoid(5,1);
F = Ellipsoid(4,1);
G = Ellipsoid(7,1);
H = Ellipsoid(8,1);
I = Ellipsoid(9,1);
J = Ellipsoid(10,1);

x_c = bias(1,1);
y_c = bias(1,2);
z_c = bias(1,3);
x_r = mag_raw/A_inv(1,1);
y_r = mag_raw/A_inv(2,2);
z_r = mag_raw/A_inv(3,3);
[j,k,l] = ellipsoid(x_c,y_c,z_c,x_r,y_r,z_r,50);

% x_c = ((2*D)*(2*H)-2*B*(2*G))/(4*A*B-(2*D)^2);
% y_c = ((2*D)*(2*G)-2*A*(2*H))/(4*A*B-(2*D)^2)
% y_c2 = ((2*F)*(2*I)-2*C*(2*H))/(4*B*C-(2*F)^2)
% z_c = ((2*F)*(2*H)-2*B*(2*I))/(4*B*C-(2*F)^2);

% myColors = ones(1, length(x));
myColors(1,length(x)) = 1; 
myColors2(1,length(x_cal)) = 0;
% change the "2" to other number to change color:
% -1:yellow; -2:magenta; 0:green; 1:cyan; 2:blue
% http://stackoverflow.com/questions/18646879/matlab-using-scatter3-and-mesh-at-the-same-time
x(isnan(myColors)) = NaN;
x_cal(isnan(myColors2)) = NaN;
[a,b,c] = sphere(50); % 50 is the mesh density

figure(1)
scatter3(x,y,z,10, myColors, 'filled'); % "100" is the size of the bubble
get(gcf, 'Renderer');
hold on;
surf(a*mag_raw+x_bias, b*mag_raw+y_bias, c*mag_raw+z_bias, 'FaceColor', [1 0 0],'EdgeAlpha',0.2,'FaceAlpha', 0.1);
hold off;
legend('Uncalibrated Results','Local Magnetic Field');

figure(2)
scatter3(x_cal,y_cal,z_cal,10, myColors2, 'filled');
hold on;
surf(a*mag_raw, b*mag_raw, c*mag_raw, 'FaceColor', [1 0 0],'EdgeAlpha',0.2,'FaceAlpha', 0.1);
hold off;
legend('Calibrated Results','Local Magnetic Field');

figure(3)
surf(j,k,l,'FaceColor', [1 1 0],'EdgeAlpha',0.2,'FaceAlpha', 0.2);
hold on
surf(a*mag_raw, b*mag_raw, c*mag_raw, 'FaceColor', [1 0 0],'EdgeAlpha',0.2,'FaceAlpha', 0.2);
hold off
legend('Uncalibrated Results (Ellipsoid)','Calibrated Results (Sphere)');

File: tools/Python_Scripts/Static_Calibration.py
#!/usr/bin/env python

"""
This script enables static data processing of Magnetometer data

Date Modified   Modified by     Changes
Feb 11, 2017    ChrisZL         Add some instrutions on the purpose of this program;
                                minor changes on the output of "_data_stats" file
                                it will now print out the combined bias and A_inv on the terminal directly
Feb 13, 2017	ChrisZL		Update the gain of the sensor

"""

from mpl_toolkits.mplot3d import Axes3D
import matplotlib.pyplot as plt
import numpy as np
import numpy.matlib
import scipy.linalg as scla
import argparse
import re


""""
 Name:      static_analysis

 Purpose:   Plot calibrated and uncalibrated data sets as well as evaluate the mean and variance

 Param:     raw and cal data
            file_name - file name to store plots

"""
def static_analysis(x_raw, y_raw, z_raw, x_cal, y_cal, z_cal, A_inv, x_bias, y_bias, z_bias, file_name, mag_raw):

    # To plot uncalibrated data
    raw_fig = plt.figure()
    ax1 = raw_fig.gca(projection='3d')

    # To plot calibrated data
    cal_fig = plt.figure()
    ax2 = cal_fig.gca(projection='3d')

    # Data histogram
    diff_fig = plt.figure()
    ax3 = diff_fig.gca()

    ax1.plot_wireframe(x_raw, y_raw, z_raw, color ='r')
    ax1.set_xlabel('x')
    ax1.set_ylabel('y')
    ax1.set_zlabel('z')

    ax2.plot_wireframe(x_cal, y_cal, z_cal, color='g')
    ax2.set_xlabel('x')
    ax2.set_ylabel('y')
    ax2.set_zlabel('z')

    diff = (numpy.sqrt(numpy.square(x_cal) + numpy.square(y_cal) + numpy.square(z_cal))) - mag_raw

    #f_stat = open("../data/%s_stats.txt" % file_name, 'w')
    #f_stat.write("Bias:\n x_bias = %.4f\n y_bias = %.4f\n z_bias = %.4f\n\n" % (x_bias, y_bias, z_bias))
    #f_stat.write("Correction matrix (A_inv):\n%s\n%s\n%s\n\n" % (A_inv[0], A_inv[1], A_inv[2]))
    #f_stat.write("Other Statisstics:\n Mean: %.4f\n Var:  %.4f\n Std:  %.4f" % (numpy.mean(diff), numpy.var(diff), numpy.std(diff)))
    #f_stat.close()

    f_stat = open("../../data/%s_Bias_&_A_inv.txt" % file_name, 'w')
    f_stat.write("%s  %s  %s\n\n" % (x_bias, y_bias, z_bias))
    f_stat.write("%f  %f  %f\n%f  %f  %f\n%f  %f  %f\n\n" % (A_inv[0,0], A_inv[0,1], A_inv[0,2], A_inv[1,0], A_inv[1,1], A_inv[1,2], A_inv[2,0], A_inv[2,1], A_inv[2,2]))
    f_stat.write("%s  %s  %s" % (numpy.mean(diff), numpy.var(diff), numpy.std(diff)))
    f_stat.close()


    hist, bins = np.histogram(diff, bins=50)
    width = 0.7 * (bins[1] - bins[0])
    center = (bins[:-1] + bins[1:]) / 2
    ax3.bar(center, hist, align='center', width=width)

    raw_fig.savefig("../../data/%s_uncal.png" % file_name)
    cal_fig.savefig("../../data/%s_cal.png" % file_name)
    diff_fig.savefig("../../data/%s_hist.png" % file_name)
    # plt.show()


"""
 Name:      static_cal

 Purpose:   Calculate A inverse matrix, bias and calibrate the magnetometer data set

 Param:     sample_path - Path to magnetometer data text file

"""
def static_cal(sample_path):
    mag_raw = 367.5 # Kelowna's raw magnetic field

    # Load data set and parse file_name from sample_path
    data = np.loadtxt(sample_path)
    f_st = re.search('/data/', sample_path)
    f_end = re.search('.txt', sample_path)
    file_name = sample_path[f_st.end():f_end.start()]

    # Write raw data to text file
    gain =  0.667
    raw_data = data* gain
    np.savetxt("../../data/%s_raw.txt" % file_name, raw_data, '%.5f')
    column_num = raw_data.shape[0]

    D = np.ones((10, column_num))
    D[0,:] = np.square(raw_data[:, 0]) # x ^ 2
    D[1,:] = np.square(raw_data[:, 1]) # y ^ 2
    D[2,:] = np.square(raw_data[:, 2]) # z ^ 2
    D[3,:] = 2 * raw_data[:, 1] * raw_data[:, 2] # 2yz
    D[4,:] = 2 * raw_data[:, 0] * raw_data[:, 2] # 2xz
    D[5,:] = 2 * raw_data[:, 0] * raw_data[:, 1] # 2xy
    D[6,:] = 2 * raw_data[:, 0] # 2x
    D[7,:] = 2 * raw_data[:, 1] # 2y
    D[8,:] = 2 * raw_data[:, 2] # 2z
    D[9,:] = np.ones((1, column_num)) #1

    # S = D * DT
    S = np.dot(D, D.transpose())

    # Split S matrix to 4 sub-matrices and evalute the SS matrix
    S_11 = S[0:6,0:6]
    S_12 = S[0:6,6:10]
    S_21 = S_12.transpose()
    S_22 = S[6:10, 6:10]
    S_22_i = np.linalg.inv(S_22)

    # SS = S_11 - S_12 * S_22_i * S_21
    SS = S_11 - np.dot(np.dot(S_12,S_22_i), S_21)

    # Constraint Matrix C
    # C = [-1, 1, 1, 0, 0, 0;
    #       1,-1, 1, 0, 0, 0;
    #       1, 1,-1, 0, 0, 0;
    #       0, 0, 0,-4, 0, 0;
    #       0, 0, 0, 0,-4, 0;
    #       0, 0, 0, 0, 0,-4];
    # Normalized C:
    C = [[0, 0.5, 0.5, 0, 0, 0],
    [0.5, 0, 0.5, 0, 0, 0],
    [0.5, 0.5, 0, 0, 0, 0],
    [0, 0, 0, -0.25, 0, 0],
    [0, 0, 0, 0, -0.25, 0],
    [0, 0, 0, 0, 0, -0.25]]

    C_i = np.linalg.inv(C)

    # Find eigenvalue of matirix E, E * V - V * D = [0]
    E = np.dot(C, SS)
    w, v = np.linalg.eig(E)

    # Find the zero - based position of the only positive eigenvalue.
    # The associated eigenvector will be in the corresponding column of matrix
    max_eigen = w[0]
    i = 0
    for k in w:
        if k>0:
            max_eigen = k
            break
        else:
            i = i+1

    # Check the sign of V1, if it is negative, flip the sign
    V1 = v[:, i]
    if V1[0] < 0:
        V1 = -1 * V1

    V2 = np.dot(np.dot(S_22_i, S_21), V1)

    # Setup Ellipsoid matrix
    Ellipsoid = np.ones((10, 1))
    Ellipsoid[0:6,0] = V1
    Ellipsoid[6:10,0] = -1 * V2[0:5]

    # Normalize Ellipsoid matrix to prevent trivial results
    Ellipsoid = Ellipsoid/np.linalg.norm(Ellipsoid)

    # Calculate the combined bias
    Q = [[Ellipsoid[0,0],Ellipsoid[5,0],Ellipsoid[4,0]], [Ellipsoid[5,0],Ellipsoid[1,0],Ellipsoid[3,0]], [Ellipsoid[4,0],Ellipsoid[3,0],Ellipsoid[2,0]]]
    Q_i = np.linalg.inv(Q)

    U = [[Ellipsoid[6,0]],[Ellipsoid[7,0]],[Ellipsoid[8,0]]]

    B = np.dot(Q_i, U)

    # Combined bias - hard Iron offset
    x_bias = -B[0,0]
    y_bias = -B[1,0]
    z_bias = -B[2,0]

    bias = np.matlib.repmat([x_bias, y_bias, z_bias], raw_data.shape[0], 1)
    
    print"\n"
    print "#-----------------------------------------------------#"
    print "This script will conduct static calibration to the targeted sensor data file."
    print "It will also do some post-processing jobs for you."
    print "Most importantly, it will give you the combined bias and A_inv matrix."
    print "They will be used in the dynamic calibration."
    print "#-----------------------------------------------------#"
    print "Combined Bias:\n%s" %(bias[0])
    
    # Find the correction matrix A_inv
    J = Ellipsoid[9,0]
    A_inv = (scla.sqrtm(Q)) * mag_raw/(np.sqrt(np.dot(np.dot(B.transpose(), Q), B) - J))
    print"\nA_inv:\n%s\n%s\n%s" %(A_inv[0], A_inv[1], A_inv[2])

    # Finally calculate and save the calibrated data
    cal_data = (np.dot(A_inv,(raw_data - bias).transpose())).transpose()
    np.savetxt("../../data/%s_cal.txt" % file_name, cal_data,'%.5e')

    # Store data points to plot and invoke plotting function
    x_raw = raw_data[:, 0]
    y_raw = raw_data[:, 1]
    z_raw = raw_data[:, 2]

    x_cal = cal_data[:, 0]
    y_cal = cal_data[:, 1]
    z_cal = cal_data[:, 2]

    # Analyze calibrated data
    static_analysis(x_raw, y_raw, z_raw, x_cal, y_cal, z_cal, A_inv, x_bias, y_bias, z_bias, file_name, mag_raw)



"""
 Name:      main

 Purpose:   Initialize command line arguments, invoke static calibration function and plot calibrated and uncalibrated
            data

 Param:     None

"""


def main():
    parser = argparse.ArgumentParser(description='Python script for static calibration of magnetometer data')
    parser.add_argument('sample_path', type=str, help='Provide data file to process')

    args = parser.parse_args()
    sample_path = args.sample_path
    static_cal(sample_path)

if __name__ == '__main__':
    main()
